internal/webserver/webserver.go

```diff
@@ -217,40 +228,45 @@ func (n kubeFilter) namespacesHandler(writer http.ResponseWriter, request *http.
 	n.handleRequest(request, username, selector)
 }
 
-func (n kubeFilter) isNamespaceListing(fn http.HandlerFunc) http.HandlerFunc {
-	return func(writer http.ResponseWriter, request *http.Request) {
-		if req.NewHttp(request, n.usernameClaimField).IsNamespaceListing() {
-			log.V(3).Info("Handling /api/v1/namespaces")
-			fn(writer, request)
-		}
-	}
-}
-
-func (n kubeFilter) isNodeListing(fn http.HandlerFunc) http.HandlerFunc {
-	return func(writer http.ResponseWriter, request *http.Request) {
-		if req.NewHttp(request, n.usernameClaimField).IsNodeListing() {
-			log.V(3).Info("Handling /api/v1/nodes")
-			fn(writer, request)
-		}
-	}
-}
-
 func (n kubeFilter) Start(ctx context.Context) error {
 	r := mux.NewRouter()
-	r.HandleFunc("/api/v1/namespaces", func(writer http.ResponseWriter, request *http.Request) {
-		n.isNamespaceListing(n.checkJwt(n.namespacesHandler))(writer, request)
-		n.reverseProxyFunc(writer, request)
-	})
-	r.HandleFunc("/api/v1/nodes", func(writer http.ResponseWriter, request *http.Request) {
-		n.isNodeListing(n.checkJwt(n.nodesHandler))(writer, request)
-		n.reverseProxyFunc(writer, request)
-	})
+	r.StrictSlash(true)
+	r.Use(n.reverseProxyMiddleware)
+
 	r.HandleFunc("/_healthz", func(writer http.ResponseWriter, request *http.Request) {
 		writer.WriteHeader(200)
 		_, _ = writer.Write([]byte("ok"))
 	})
+
+	ns := r.Path("/api/v1/namespaces").Subrouter()
+	ns.Use(n.checkJWTMiddleware)
+	ns.HandleFunc("", func(writer http.ResponseWriter, request *http.Request) {
+		n.namespacesHandler(writer, request)
+	})
+
+	node := r.Path("/api/v1/nodes").Subrouter()
+	node.Use(n.checkJWTMiddleware)
+	node.HandleFunc("", func(writer http.ResponseWriter, request *http.Request) {
+		n.nodesHandler(writer, request)
+	})
+
 	r.PathPrefix("/").HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
-		n.reverseProxyFunc(writer, request)
+		if n.serverOptions.IsListeningTls() {
```

Comment:
Just for the catch-all route (so the ones non overloaded by our reverse-proxy) we're doing the _impersonification_: this is required since the reverse proxy inherited the transport from the cluster administrator `kubeconfig` (or the equivalent) and the following requests should be presented on behalf of the requester, and _impersonification_ is exactly doing this.